pallets/pallet-bonded-coins/src/curves/lmsr.rs

```diff
@@ -34,13 +34,30 @@ where
 	Parameter: FixedSigned + PartialOrd<Precision> + From<Precision> + ToFixed,
 	<Parameter as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
 {
-	fn calculate_exp_term(&self, supply: Parameter, x: Parameter) -> Result<Parameter, ArithmeticError> {
-		supply
-			.checked_sub(x)
-			.ok_or(ArithmeticError::Underflow)?
-			.checked_div(self.m)
-			.ok_or(ArithmeticError::DivisionByZero)
-			.and_then(|exponent| exp::<Parameter, Parameter>(exponent).map_err(|_| ArithmeticError::Overflow))
+	fn lse(&self, supply: &[Parameter]) -> Result<Parameter, ArithmeticError> {
+		// Find the maximum value in the supply for numerical stability
+		let max = supply.iter().max().ok_or_else(|| {
+			log::error!(target: LOG_TARGET, "Supply is empty. Found pool with no currencies.");
+			ArithmeticError::Underflow
+		})?;
+
+		// Compute the sum of the exponent terms, adjusted by max for stability
+		let e_term_sum = supply.iter().try_fold(Parameter::from_num(0), |acc, x| {
+			let exponent = x
+				.checked_sub(*max)
+				.ok_or(ArithmeticError::Underflow)?
+				.checked_div(self.m)
+				.ok_or(ArithmeticError::DivisionByZero)?;
+
+			let exp_result = exp::<Parameter, Parameter>(exponent).map_err(|_| ArithmeticError::Overflow)?;
+			acc.checked_add(exp_result).ok_or(ArithmeticError::Overflow)
+		})?;
+
+		// Compute the logarithm of the sum and scale it by `m`, then add the max term
+		ln::<Parameter, Parameter>(e_term_sum)
```

Comment:
`e_term_sum` is 0 when all the items in `supply` are 0, correct? This would then fail, is that a problem?

Reply:
should we just fall back to returning 0 in that case?